mmengine/runner/deepspeed_runner.py

```diff
@@ -0,0 +1,771 @@
+                        group=process_group,
+                        device=get_device(),
+                    )
+                    self._all_state_dicts.append(
+                        _recursive_copy_to_device(
+                            local_state_dict,
+                            non_blocking=True,
+                            device=torch.device('cpu')))
+            else:
+                if rank == local_rank:
+                    # Send the optimizer state to the target rank
+                    _ = _broadcast_object(
+                        state_dict,
+                        src_rank=global_rank,
+                        group=process_group,
+                        device=get_device(),
+                    )
+                elif rank != to:
+                    # Discard the received object; `broadcast()` is used for
+                    # compatibility reasons
+                    _ = _broadcast_object(
+                        empty_messenger,
+                        src_rank=global_rank,
+                        group=process_group,
+                        device=get_device(),
+                    )
+
+    def get_zero_state_dict(self) -> Dict[str, Any]:
+        r"""
+        Returns the last global optimizer state known to this rank.
+        .. warning:
+            If the state has not been consolidated to this rank, this raises a
+            runtime error, and even if it has, the state may not be up-to-date,
+            depending on when :meth:`consolidate_state_dict` was last called.
+        """
+        if not is_main_process():
+            return dict()
+
+        if len(self._all_state_dicts) == 0:
+            raise RuntimeError(
+                'Optimizer state has not been consolidated on this rank. '
+                f'Please call `consolidate_state_dict(to={self.rank})` on '
+                'all ranks beforehand if you meant to save the global state.')
+
+        return self._all_state_dicts
+
+    def get_state_dict(self, model):
+        from deepspeed.checkpoint.constants import FP32_FLAT_GROUPS
+        from deepspeed.utils.zero_to_fp32 import \
+            _get_fp32_state_dict_from_zero3_checkpoint
+
+        if not is_main_process():
+            return dict()
+
+        # ZeRO 3 case
+        if self.model.zero_optimization_partition_weights():
+            optim_state = self.get_zero_state_dict()
+            fp32_flat_groups = [
+                torch.cat(optim_state[i][FP32_FLAT_GROUPS])
+                for i in range(len(optim_state))
+            ]
+            param_shapes = self.model._get_zero_param_shapes()[0]
+            param_shapes = OrderedDict(
+                {'module.' + k: v
+                 for k, v in param_shapes.items()})
+
+            model_state = _get_fp32_state_dict_from_zero3_checkpoint(
+                world_size=self._world_size,
+                param_shapes=[param_shapes],
+                fp32_flat_groups=fp32_flat_groups,
+                buffers={})
```

Comment:
There should be special logic for ZeRO 3.
https://github.com/microsoft/DeepSpeed/blob/211055216792cbb52ab6d355f698c194f9c55efb/deepspeed/utils/zero_to_fp32.py#L100

Reply:
In the case of ZeRO3, `model_state` are saved in the `optim_state`.
https://github.com/microsoft/DeepSpeed/issues/2413